central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -80,103 +98,170 @@ func (ds *datastoreImpl) initGraph() error {
 		if err != nil {
-			deleteErr := ds.graph.DeleteVertex(obj.GetId())
-			if deleteErr != nil {
-				log.Errorf("Failed to remove collection from internal state object (%v)", deleteErr)
-			}
-			return err
+			return nil, err
 		}
 	}
-	return nil
+
+	// set id for the object and return the new graph
+	obj.Id = id
+	return graph, nil
 }
 
-func (ds *datastoreImpl) deleteCollectionFromGraph(id string) error {
+func (ds *datastoreImpl) updateCollectionInGraph(obj *storage.ResourceCollection) (*dag.DAG, error) {
 	ds.initGraphOnce()
-	ds.graphLock.Lock()
-	defer ds.graphLock.Unlock()
 
-	return ds.graph.DeleteVertex(id)
-}
+	var err error
 
-func (ds *datastoreImpl) dryRunAddCollectionInGraph(obj *storage.ResourceCollection) error {
-	ds.initGraphOnce()
-	ds.graphLock.Lock()
-	defer ds.graphLock.Unlock()
+	if err = verifyCollectionObjectNotEmpty(obj); err != nil {
+		return nil, err
+	}
 
-	var ret error
+	addChildIDs := make([]string, 0)
+	removeChildIDs := make([]string, 0)
 
-	// we essentially just add the obj to the graph and delete it before returning
-	err := ds.graph.AddVertexByID(obj.GetId(), obj.GetName())
+	// create graph copy to use for operation
+	graph, err := ds.graph.Copy()
 	if err != nil {
-		return err
+		return nil, err
+	}
+
+	// get current children edges for the object
+	curChildren, err := graph.GetChildren(obj.GetId())
+	if err != nil {
+		return nil, errors.Wrap(err, "update attempt on unknown collection")
+	}
+
+	// determine additions
+	for _, newChild := range obj.GetEmbeddedCollections() {
+		_, present := curChildren[newChild.GetId()]
+		if present {
+			// if present in both we remove from curChildren
+			delete(curChildren, newChild.GetId())
+		} else {
+			addChildIDs = append(addChildIDs, newChild.GetId())
+		}
+	}
+
+	// determine deletions, any remaining edge in curChildren should be removed
+	for key := range curChildren {
+		removeChildIDs = append(removeChildIDs, key)
 	}
-	for _, peer := range obj.GetEmbeddedCollections() {
-		err = ds.graph.AddEdge(obj.GetId(), peer.GetId())
+
+	// since adding can cause cycles but removing never can, we remove first
+	for _, childID := range removeChildIDs {
+		err = graph.DeleteEdge(obj.GetId(), childID)
 		if err != nil {
-			ret = err
-			break
+			return nil, err
 		}
 	}
 
-	err = ds.graph.DeleteVertex(obj.GetId())
-	if err != nil && ret == nil {
-		ret = err
+	// add new edges
+	for _, childID := range addChildIDs {
+		err = graph.AddEdge(obj.GetId(), childID)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	return ret
+	return graph, nil
+}
+
+func (ds *datastoreImpl) deleteCollectionFromGraph(id string) error {
+	ds.initGraphOnce()
+
+	// this function covers removal of the vertex and any edges to or from other vertices, it is also threadsafe
+	return ds.graph.DeleteVertex(id)
 }
 
 func (ds *datastoreImpl) Search(ctx context.Context, q *v1.Query) ([]pkgSearch.Result, error) {
+	ds.lock.RLock()
+	defer ds.lock.RUnlock()
```

Comment:
locking is not required because neither the graph nor the name set is accessed.

Reply:
Do we not have to worry about concurrency at the postgres level?

Reply:
`SELECT` queries are implicitly wrapped in a transaction and work on the last committed state.